### DISCUSSION TITLE
NODE_ENV not set initially prior to configValidation for 'npm run build'

### DISCUSSION THREAD

**Discussion by miller-productions** [facebook/docusaurus#8549] [Q&A] [answered]
Hi folks,

I'm trying out using dotenv to dynamically set different values for `url` in the file `docusaurus.config.js`

For example...

In `docusaurus.config.js`, I have the following relevant lines...

```
// load environment variables from relavent .env file
require('dotenv').config({ path: `.env.${process.env.NODE_ENV}` })
...etc
const config = {
  title: 'my title',
  tagline: 'my tagline',
  url: process.env.DOCS_URL,
etc...
```

Then, I have separate dotenv files as such

```
// .env.development
DOCS_URL=http://localhost:3000
```
and
```
// .env.production
DOCS_URL=https://my-docs.someurl
```

Now, with this config setup, if I run `npm start`, everything works fine. NODE_ENV gets set to `development`, the relevant dotenv file gets loaded, all is good.

However, if I run `npm run build`, no joy. A quick check of `console.log(process.env.NODE_ENV);` inside `docusaurus.config.js` reveals that in this context, NODE_ENV is `undefined`.

Would anyone be able to assist? Am I missing something here?

-------------

Further info.

I made the following second attempt to check something...

I added a fallback value for that `url` config setting in case the dotenv config value can't be evaluated.

```
// load environment variables from relavent .env file
require('dotenv').config({ path: `.env.${process.env.NODE_ENV}` })
console.log("process.env.NODE_ENV is: ", process.env.NODE_ENV);
...etc
const config = {
  title: 'my title',
  tagline: 'my tagline',
  url: process.env.DOCS_URL || 'https://my-docs.someurl',
etc...
```

and I noticed that `docusaurus.config.js` appears to load twice, resulting in the following console output...

```
$ npm run build

> docs-website@0.0.0 build
> docusaurus build

process.env.NODE_ENV is:  undefined
[INFO] [en] Creating an optimized production build...
process.env.NODE_ENV is:  production

✔ Client
  

✔ Server
  Compiled successfully in 18.64s


✔ Client
  

● Server █████████████████████████ cache (99%) shutdown IdleFileCachePlugin
 stored

[SUCCESS] Generated static files in "build".
[INFO] Use `npm run serve` command to test your build locally.
```
So it looks like the first time through, `process.env.NODE_ENV` isn't set, which result in an error in `node_modules\@docusaurus\core\lib\server\configValidation.js` i.e. `[ERROR] Error: "url" is required` if you only rely on dotenv config alone, without the fallback as per my second attempt

Do you think this is a bug worth reporting?

**Comment by nicocodemance** [comment, marked as answer]
@miller-productions  Try this one

```js
require('dotenv').config()
const config = {
  title: 'my title',
  tagline: 'my tagline',
  url: process.env.DOCS_URL || 'https://my-docs.someurl',
}
```

It will work for you

**Reply by miller-productions** [reply]
Thanks @nicholidev

**Reply by purecodecoach** [reply]
Is it working?

**Reply by miller-productions** [reply]
With the fallback provided to url, i.e. this... `url: process.env.DOCS_URL || 'https://my-docs.someurl',` then, yes, it is a sufficient workaround.

**Comment by Josh-Cena** [comment]
@miller-productions Indeed, we load the config once before we even set `NODE_ENV`, mainly to figure out the locales we need to build. However, you should not rely on `NODE_ENV` being not undefined, because in some commands (e.g. `swizzle`), the env is not set throughout the process.

@slorber Do you think we should make sure every command has a `NODE_ENV` set in the entire process? Could be helpful for `.env` loading.

**Comment by Josh-Cena** [comment]
Created issue at https://github.com/facebook/docusaurus/issues/8599
